hazelcast/src/main/java/com/hazelcast/spi/impl/classicscheduler/ClassicOperationScheduler.java

```diff
@@ -0,0 +1,432 @@
+                    threadName, threadId, genericWorkQueue, genericPriorityWorkQueue,
+                    logger, threadGroup, nodeExtension, operationHandler);
+            threads[threadId] = operationThread;
+            operationThread.start();
+        }
+
+        return threads;
+    }
+
+    @Override
+    public OperationHandler[] getPartitionOperationHandlers() {
+        return partitionOperationHandlers;
+    }
+
+    @Override
+    public OperationHandler[] getGenericOperationHandlers() {
+        return genericOperationHandlers;
+    }
+
+    @Override
+    public boolean isAllowedToRunInCurrentThread(Operation op) {
+        return isAllowedToRunInCurrentThread(op.getPartitionId());
+    }
+
+    @Override
+    public boolean isOperationThread() {
+        return Thread.currentThread() instanceof OperationThread;
+    }
+
+    @Override
+    public boolean isInvocationAllowedFromCurrentThread(Operation op) {
+        return isInvocationAllowedFromCurrentThread(op.getPartitionId());
+    }
+
+    private boolean isAllowedToRunInCurrentThread(int partitionId) {
+        Thread currentThread = Thread.currentThread();
+
+        // IO threads are not allowed to run any operation
+        if (currentThread instanceof NIOThread) {
+            return false;
+        }
+
+        //todo: do we want to allow non partition specific tasks to be run on a partitionSpecific operation thread?
+        if (partitionId < 0) {
+            return true;
+        }
+
+        //we are only allowed to execute partition aware actions on an OperationThread.
+        if (!(currentThread instanceof PartitionOperationThread)) {
+            return false;
+        }
+
+        PartitionOperationThread operationThread = (PartitionOperationThread) currentThread;
+
+        //so it is an partition operation thread, now we need to make sure that this operation thread is allowed
+        //to execute operations for this particular partitionId.
+        return toPartitionThreadIndex(partitionId) == operationThread.threadId;
+    }
+
+    private boolean isInvocationAllowedFromCurrentThread(int partitionId) {
+        Thread currentThread = Thread.currentThread();
+
+        if (currentThread instanceof PartitionOperationThread) {
+            PartitionOperationThread operationThread = (PartitionOperationThread) currentThread;
+            return toPartitionThreadIndex(partitionId) == operationThread.threadId;
+        }
+
+        // IO threads are not allowed to run any operation
+        if (currentThread instanceof NIOThread) {
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public int getRunningOperationCount() {
+        int result = 0;
+        for (PartitionOperationThread thread : partitionOperationThreads) {
+            if (thread.currentOperation != null) {
+                result++;
+            }
+        }
+        for (OperationThread thread : genericOperationThreads) {
```

Comment:
I would use `OperationThread` in Line 266, too or `GenericOperationThread` here. The mix confused me, since I thought there was no `GenericOperationThread` at all (from this code snippet).